lib/elixir/lib/dynamic_supervisor.ex

```diff
@@ -0,0 +1,859 @@
+
+  Note that calling this function when supervising a large number
+  of children under low memory conditions can cause an out of memory
+  exception.
+
+  This function returns a list of tuples containing:
+
+    * `id` - as defined in the child specification but is always
+      set to `:undefined` for dynamic supervisors
+
+    * `child` - the pid of the corresponding child process or the
+      atom `:restarting` if the process is about to be restarted
+
+    * `type` - `:worker` or `:supervisor` as defined in the child
+      specification
+
+    * `modules` - as defined in the child specification
+
+  """
+  @spec which_children(Supervisor.supervisor()) :: [
+          {:undefined, pid | :restarting, :worker | :supervisor, :supervisor.modules()}
+        ]
+  def which_children(supervisor) do
+    call(supervisor, :which_children)
+  end
+
+  @doc """
+  Returns a map containing count values for the supervisor.
+
+  The map contains the following keys:
+
+    * `:specs` - always 1 as dynamic supervisors have a single specification
+
+    * `:active` - the count of all actively running child processes managed by
+      this supervisor
+
+    * `:supervisors` - the count of all supervisors whether or not the child
+      process is still alive
+
+    * `:workers` - the count of all workers, whether or not the child process
+      is still alive
+
+  """
+  @spec count_children(Supervisor.supervisor()) :: %{
+          specs: non_neg_integer,
+          active: non_neg_integer,
+          supervisors: non_neg_integer,
+          workers: non_neg_integer
+        }
+  def count_children(supervisor) do
+    call(supervisor, :count_children)
+  end
+
+  @doc """
+  Receives a set of options that initializes a dynamic supervisor.
+
+  This is typically invoked at the end of the `c:init/1` callback of
+  module-based supervisors. See the sections "Module-based supervisors"
+  in the module documentation for more information.
+
+  The options received by this function are also supported by `start_link/2`.
+
+  This function returns a tuple containing the supervisor options.
+
+  ## Examples
+
+      def init(_arg) do
+        DynamicSupervisor.init(max_children: 1000, strategy: :one_for_one)
+      end
+
+  ## Options
+
+    * `:strategy` - the restart strategy option. The only supported
+      value is `:one_for_one` which means that no other child is
+      terminate if a child process terminates. You can learn more
+      about strategies in the `Supervisor` module docs.
+
+    * `:max_restarts` - the maximum number of restarts allowed in
+      a time frame. Defaults to `3`.
+
+    * `:max_seconds` - the time frame in which `:max_restarts` applies.
+      Defaults to `5`.
+
+    * `:max_children` - the maximum amount of children to be running
+      under this supervisor at the same time. When `:max_children` is
+      exceeded, `start_child/2` returns `{:error, :dynamic}`.
+
+  """
+  def init(options) when is_list(options) do
```

Comment:
As mentioned, I don't see why do we need to invoke this function ourselves. If you do decide to keep that approach, then I think it should be specced.